service/src/test/java/bio/terra/workspace/service/iam/ControlledResourceSamPolicyBuilderTest.java

```diff
@@ -0,0 +1,180 @@
+        equalTo(false));
+    assertThat(
+        request.getPolicies().containsKey(ControlledResourceIamRole.READER.toSamRole()),
+        equalTo(false));
+    assertThat(
+        request.getPolicies().containsKey(ControlledResourceIamRole.DELETER.toSamRole()),
+        equalTo(false));
+  }
+
+  @Test
+  void addPolicies_applicationShared() {
+    var app =
+        new WsmWorkspaceApplication()
+            .application(new WsmApplication().serviceAccount("fake-svc@example.com"));
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            null, null, ControlledResourceCategory.APPLICATION_SHARED, app);
+    var request = new CreateResourceRequestV2();
+
+    policyBuilder.addPolicies(request);
+
+    assertThat(
+        request.getPolicies().get(ControlledResourceIamRole.OWNER.toSamRole()).getMemberEmails(),
+        contains(GcpUtils.getWsmSaEmail()));
+    assertThat(
+        request.getPolicies().get(ControlledResourceIamRole.EDITOR.toSamRole()).getMemberEmails(),
+        contains("fake-svc@example.com"));
+  }
+
+  @Test
+  void addPolicies_applicationSharedFailsWithPrivateIamRole() {
+    var app =
+        new WsmWorkspaceApplication()
+            .application(new WsmApplication().serviceAccount("fake-svc@example.com"));
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            ControlledResourceIamRole.EDITOR,
+            null,
+            ControlledResourceCategory.APPLICATION_SHARED,
+            app);
+    var request = new CreateResourceRequestV2();
+
+    assertThrows(
+        InternalLogicException.class,
+        () -> policyBuilder.addPolicies(request),
+        "Specifying a private IAM role on an application shared resource is invalid");
+  }
+
+  @Test
+  void addPolicies_applicationSharedWithNoAppFails() {
+    // ensure an application shared resource cannot have a user with a private IAM role on it
+    // (tho choice of EDITOR for the private IAM role is arbitrary--any value will cause an
+    // exception.
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            ControlledResourceIamRole.EDITOR,
+            null,
+            ControlledResourceCategory.APPLICATION_SHARED,
+            null);
+    var request = new CreateResourceRequestV2();
+
+    assertThrows(InternalLogicException.class, () -> policyBuilder.addPolicies(request));
+  }
+
+  @Test
+  void addPolicies_applicationSharedWithUserEmailFails() {
+    var app =
+        new WsmWorkspaceApplication()
+            .application(new WsmApplication().serviceAccount("fake-svc@example.com"));
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            ControlledResourceIamRole.EDITOR,
+            "fake@example.com",
+            ControlledResourceCategory.APPLICATION_SHARED,
+            app);
+    var request = new CreateResourceRequestV2();
+
+    assertThrows(InternalLogicException.class, () -> policyBuilder.addPolicies(request));
```

Comment:
Since an InternalLogicException can be thrown for a couple of different reasons, it would be good to check the error message.